internal/ioutil/rereader.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ioutil
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	"go.uber.org/yarpc/internal/buffer"
+
+	"github.com/uber-go/atomic"
+)
+
+// TODO Optimizations:
+// - Attempt to cast the source reader to a *bytes.Buffer and use it directly
+//   instead of needing to copy into a new buffer.
+// - In cases where the first reader was not completely read when we call
+//   reset we should flush the rest of the reader into the buffer, or
+//   combine the buffer and source reader together dynamically.
+
+// NewRereader returns a new re-reader and closer that wraps the given reader.
+// The re-reader consumes the given reader on demand, recording the entire
+// stream so it can be replayed. The re-reader is suitable for retries. The
+// re-reader does not support concurrent consumers, as would be necessary
+// for speculative retries.
+func NewRereader(src io.Reader) (*ReReader, func()) {
```

Comment:
Can't hide the Rereader unfortunately, we need the Reset function.  People are forced to use the constructor because all of the Rereader fields are private, so they can create one, but it won't work at all.